prometheus/counter.go

```diff
@@ -129,11 +129,22 @@ func (c *counter) Add(v float64) {
 	}
 
 	ival := uint64(v)
-	if float64(ival) == v {
-		atomic.AddUint64(&c.valInt, ival)
-		return
-	}
 
+	// if the v is an unsigned integer
+	// and the precise doesn't lose during uint cast and cast it back to float
+	for v == float64(ival) {
```

Comment:
one thing to note here is that the casting here varies in different platform. for example, in linux, the v won't be equal with `float64(uint64(v))`. but in macos m1, it equals.

The verification code is:
```go
func TestVerifyRounding(t *testing.T) {
	var f float64 = 1<<64 - 1
	if f == float64(uint64(f)) {
		t.Fatal("values are the same after casting")
	} else {
		t.Fatal("values are different")
	}
}
```
LInux output could be found in the [pipeline here](https://github.com/xieyuschen/client_golang/actions/runs/8339696721/job/22822165848).
![image](https://github.com/prometheus/client_golang/assets/52945328/08e5512d-1f92-4826-b246-e6e1af5d3634)

and the macos outputs:
![image](https://github.com/prometheus/client_golang/assets/52945328/839cc6a1-7a6b-47a7-a606-08d3ae070999)